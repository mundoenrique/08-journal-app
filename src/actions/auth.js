import {
	createUserWithEmailAndPassword,
	getAuth,
	signInWithEmailAndPassword,
	signInWithPopup,
	updateProfile,
} from 'firebase/auth';

import { googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from '../types/types';

export const startLoginEmailPassword = (email, password) => {
	return (dispatch) => {
		const auth = getAuth();
		signInWithEmailAndPassword(auth, email, password)
			.then(({ user }) => {
				console.log(user);
				dispatch(login(user.uid, user.displayName));
			})
			.catch((error) => {
				console.log(typeof error);
				console.log(error);
				console.log(error.code);
				console.log(error.message);
			});
	};
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
	return (dispatch) => {
		const auth = getAuth();
		createUserWithEmailAndPassword(auth, email, password)
			.then(async ({ user }) => {
				await updateProfile(user, { displayName: name });
				console.log(user);
				dispatch(login(user.uid, user.displayName));
			})
			.catch((error) => {
				console.log(typeof error);
				console.log(error);
				console.log(error.code);
				console.log(error.message);
			});
	};
};

export const startGoogleLogin = () => {
	return (dispatch) => {
		const auth = getAuth();
		signInWithPopup(auth, googleAuthProvider)
			.then(({ user }) => {
				dispatch(login(user.uid, user.displayName));
			})
			.catch((error) => {
				console.log(error);
			});
	};
};

export const login = (uid, displayName) => ({
	type: types.login,
	payload: {
		uid,
		displayName,
	},
});
