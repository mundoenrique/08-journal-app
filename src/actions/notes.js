import Swal from 'sweetalert2';
import {
	addDoc,
	collection,
	deleteDoc,
	doc,
	getDocs,
	setDoc,
} from 'firebase/firestore';

import { db } from '../firebase/firebaseConfig';
import { types } from '../types/types';
import { fileUpload } from '../helpers/fileUpload';

export const startAddNewNote = () => {
	return async (dispatch, getState) => {
		const { uid } = getState().auth;

		const newNote = {
			title: '',
			body: '',
			date: new Date().getTime(),
		};

		const docRef = await addDoc(
			collection(db, uid, 'journal', 'notes'),
			newNote
		);

		dispatch(activeNote(docRef.id, newNote));
		dispatch(addNewNote(docRef.id, newNote));
	};
};

export const activeNote = (id, note) => ({
	type: types.notesActive,
	payload: {
		id,
		...note,
	},
});

export const addNewNote = (id, note) => ({
	type: types.notesAddNew,
	payload: {
		id,
		...note,
	},
});

export const startLoadNotes = () => {
	return async (dispatch, getState) => {
		const { uid } = getState().auth;

		const notes = [];
		const querySnapshot = await getDocs(collection(db, `${uid}/journal/notes`));

		querySnapshot.forEach((doc) => {
			notes.push({
				id: doc.id,
				...doc.data(),
			});
		});

		dispatch(setNotes(notes));
	};
};

export const setNotes = (notes) => ({
	type: types.notesLoad,
	payload: notes,
});

export const startSaveNote = (note) => {
	return async (dispatch, getState) => {
		const { uid } = getState().auth;

		if (!note.url) {
			delete note.url;
		}

		const noteToFirestore = { ...note };
		delete noteToFirestore.id;

		await setDoc(doc(db, uid, 'journal', 'notes', note.id), noteToFirestore);

		dispatch(refreshNote(note.id, noteToFirestore));
		Swal.fire('Saved!', note.title, 'success');
	};
};

export const refreshNote = (id, note) => ({
	type: types.notesUpdated,
	payload: { id, note: { id, ...note } },
});

export const startUploading = (file) => {
	return async (dispatch, getState) => {
		const { active: note } = getState().notes;

		Swal.fire({
			title: 'Uploading...',
			html: 'Please wait...',
			allowOutsideClick: false,
			didOpen: () => {
				Swal.showLoading();
			},
		});

		const fileUrl = await fileUpload(file);
		note.url = fileUrl;

		dispatch(startSaveNote(note));

		Swal.close();
	};
};

export const startDeleteNote = (id) => {
	return async (dispatch, getState) => {
		const { uid } = getState().auth;

		await deleteDoc(doc(db, uid, 'journal', 'notes', id));

		dispatch(deleteNote(id));
	};
};

export const deleteNote = (id) => ({
	type: types.notesDelete,
	payload: id,
});

export const noteLogout = () => ({
	type: types.notesLogoutCleaning,
});
