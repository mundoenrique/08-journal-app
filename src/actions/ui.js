import { types } from '../types/types';

export const SetError = (err) => ({
	type: types.uiSetError,
	payload: err,
});

export const removeError = () => ({
	type: types.uiRemoveError,
});
