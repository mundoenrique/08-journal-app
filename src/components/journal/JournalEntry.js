export default function JournalEntry({ entry }) {
	return (
		<>
			<div className="journal__entry pointer">
				<div
					className="journal__entry-picture"
					style={{
						backgroundSize: 'cover',
						backgroundImage: `url(https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Image_created_with_a_mobile_phone.png/640px-Image_created_with_a_mobile_phone.png)`,
					}}
				></div>
				<div className="journal__entry-body">
					<p className="journal__entry-title">Un nuevo d√≠a</p>
					<p className="journal__entry-content">
						Lorem Ipsum is simply dummy text of the printing and typesetting
						industry.
					</p>
				</div>
				<div className="journal__entry-date-box">
					<span>Lunes</span>
					<h4>28</h4>
				</div>
			</div>
		</>
	);
}
