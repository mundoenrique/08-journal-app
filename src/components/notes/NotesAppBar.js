import { useDispatch, useSelector } from 'react-redux';
import { startSaveNote } from '../../actions/notes';

export default function NotesAppBar() {
	const dispatch = useDispatch();
	const { active: note } = useSelector((state) => state.notes);

	const handlesave = () => {
		dispatch(startSaveNote(note));
	};

	return (
		<>
			<div className="notes__appbar">
				<span>28 de Agosto 2020</span>
				<div>
					<button className="btn">Picture</button>
					<button className="btn" onClick={handlesave}>
						Save
					</button>
				</div>
			</div>
		</>
	);
}
