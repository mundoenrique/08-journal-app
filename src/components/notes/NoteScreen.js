import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import NotesAppBar from './NotesAppBar';
import { activeNote, startDeleteNote } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';

export default function NoteScreen() {
	const dispatch = useDispatch();
	const { active: note } = useSelector((state) => state.notes);
	const [inputValues, handleInputValues, resetValues] = useForm(note);
	const { title, body, id } = inputValues;
	const activeId = useRef(note.id);

	useEffect(() => {
		if (note.id !== activeId.current) {
			resetValues(note);
			activeId.current = note.id;
		}
	}, [note, resetValues]);

	useEffect(() => {
		dispatch(activeNote(inputValues.id, { ...inputValues }));
	}, [dispatch, inputValues]);

	const handleDelete = () => {
		dispatch(startDeleteNote(id));
	};

	return (
		<>
			<div className="notes__main-content">
				<NotesAppBar />
				<div className="notes__content">
					<input
						type="text"
						placeholder="Some awesome title"
						className="notes__title-input"
						name="title"
						autoComplete="off"
						value={title}
						onChange={handleInputValues}
					/>
					<textarea
						placeholder="What happened today"
						className="notes__textarea"
						name="body"
						value={body}
						onChange={handleInputValues}
					></textarea>
					{note.url && (
						<div className="notes__image">
							<img src={note.url} alt={title} />
						</div>
					)}
				</div>
				<button className="btn btn-danger" onClick={handleDelete}>
					Delete
				</button>
			</div>
		</>
	);
}
