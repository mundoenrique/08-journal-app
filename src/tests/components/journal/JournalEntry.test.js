import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

import JournalEntry from '../../../components/journal/JournalEntry';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};

let store = mockStore(initState);
store.dispatch = jest.fn();

const note = {
	id: '123',
	title: 'Note title',
	body: 'Note body',
	date: 5646546546132131,
	url: 'https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png',
};

const wrapper = mount(
	<Provider store={store}>
		<JournalEntry {...note} />
	</Provider>
);

describe('Pruebas en <JournalEntry />', () => {
	test('Debe renderizar correctament <JournalEntry />', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('Debe llamar al activenote', () => {
		wrapper.find('.journal__entry').prop('onClick')();

		expect(store.dispatch).toHaveBeenCalled();
		expect(store.dispatch).toHaveBeenCalledWith(
			activeNote(note.id, { ...note })
		);
	});
});
