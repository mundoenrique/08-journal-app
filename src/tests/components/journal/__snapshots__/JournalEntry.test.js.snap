// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <JournalEntry /> Debe renderizar correctament <JournalEntry /> 1`] = `
<div
  className="journal__entry pointer  animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://www.google.com/images/branding/googlelogo/2x/googlelogo_color_272x92dp.png)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Note title
    </p>
    <p
      className="journal__entry-content"
    >
      Note body
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Thursday
    </span>
    <h4>
      8th
    </h4>
  </div>
</div>
`;
